datasets/noaa/_images/run_csv_transform_kub_gsod_stations/csv_transform.py

```diff
@@ -0,0 +1,304 @@
+        return val
+    else:
+        if val.find("\n") > 0:
+            # return val.replace("\n", "")
+            return re.sub(r"(^\d):(\d{2}:\d{2})", "0$1:$2", val)
+        else:
+            return val
+
+
+def replace_values_regex(df: pd.DataFrame) -> None:
+    header_names = {"checkout_time"}
+
+    for dt_col in header_names:
+        if df[dt_col] is not None:
+            if df[dt_col].str.len() > 0:
+                df[dt_col] = df[dt_col].apply(replace_value)
+
+
+def download_file_ftp(
+    ftp_host: str,
+    ftp_dir: str,
+    ftp_filename: str,
+    local_file: pathlib.Path,
+    source_url: str,
+) -> None:
+
+    # ftp_host      -> host ip (eg. 123.123.0.1)
+    # ftp_dir       -> working directory in ftp host where file is located
+    # ftp_filename  -> filename of FTP file to download
+    # source_file   -> local file (including path) to create containing ftp content
+
+    logging.info(f"Downloading {source_url} into {local_file}")
+    ftp_conn = FTP(ftp_host)
+    ftp_conn.login("", "")
+    ftp_conn.cwd(ftp_dir)
+
+    try:
+        bak_local_file = str(local_file) + ".bak"
+        dest_file = open(bak_local_file, "wb")
+        ftp_conn.encoding = "utf-8"
+        ftp_conn.retrbinary(
+            cmd="RETR " + ftp_filename,
+            callback=dest_file.write,
+            blocksize=1024,
+            rest=None,
+        )
+        ftp_conn.quit()
+        dest_file.close()
+    except Exception as e:
+        logging.error(f"Error saving output file: {e}.")
+
+
+def gz_decompress(infile: str, tofile: str) -> None:
+    # open infile (gzip file) for read and opfile for write (decompressed)
+    with open(infile, "rb") as inf, open(tofile, "w", encoding="utf8") as tof:
+        # decompress the file
+        decom_str = gzip.decompress(inf.read()).decode("utf-8")
+        # write the file as decompressed
+        tof.write(decom_str)
+
+
+def url_is_reachable(url: str) -> None:
+    # Is the URL reachable?
+
+    request = urllib.request.Request(url)
+    request.get_method = lambda: "HEAD"
+
+    try:
+        urllib.request.urlopen(request)
+        return True
+    except urllib.request.HTTPError:
+        return False
+
+
+def download_file(source_url: str, source_file: pathlib.Path) -> None:
+    r = requests.get(source_url, stream=True)
+    if r.status_code == 200:
+        with open(source_file, "wb") as f:
+            for chunk in r:
+                f.write(chunk)
+    else:
+        logging.error(f"Couldn't download {source_url}: {r.text}")
+
+
+def upload_file_to_gcs(file_path: pathlib.Path, gcs_bucket: str, gcs_path: str) -> None:
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(gcs_bucket)
+    blob = bucket.blob(gcs_path)
+    blob.upload_from_filename(file_path)
+
+
+if __name__ == "__main__":
+    logging.getLogger().setLevel(logging.INFO)
+
+    # source url should contain the year as {year} in the source url eg. "https://my.site.gov/../../data_file_{year}.csv.gz"
```

Comment:
Up to you, whether you'd like to delete it, or move it to the YAML config as a comment / rule in the input value.

Reply:
Done